### DISCUSSION TITLE
How to get the git commit SHA with Next.js?

### DISCUSSION THREAD

**Discussion by Vadorequest** [vercel/community#56] [Help] [answered]
I tried various attempts to know the git commit sha/hash at build time.
Unfortunately, none works.

How can I know the git SHA at build time? (or runtime)

```js
console.log('process.env.VERCEL_GIT_COMMIT_SHA: ', process.env.VERCEL_GIT_COMMIT_SHA);
console.log('process.env.NEXT_PUBLIC_VERCEL_GIT_COMMIT_SHA: ', process.env.NEXT_PUBLIC_VERCEL_GIT_COMMIT_SHA);
console.log('process.env.GIT_COMMIT_SHA: ', process.env.GIT_COMMIT_SHA);
```

![image](https://user-images.githubusercontent.com/3807458/102395521-8a8fce00-3fdb-11eb-9750-9062cdc88ff0.png)

**Comment by styfle** [comment, marked as answer]
If you want to use the the value on the client, then it will need to be prefixed with [`NEXT_PUBLIC_`](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser).

**Update**: Vercel automatically prefixes [System Environment Variables](https://vercel.com/docs/projects/environment-variables/system-environment-variables) so you can use `VERCEL_GIT_COMMIT_SHA` on the backend and use `NEXT_PUBLIC_VERCEL_GIT_COMMIT_SHA` if you want to expose to the client.

**Reply by Vadorequest** [reply]
Does this works when not using the native GitHub integration? I don't think so. (ran some experiment and got those empty)

**Reply by styfle** [reply]
You have to use either GitHub/GitLab/Bitbucket integration. [See docs](https://vercel.com/docs/git)

Otherwise Vercel doesn't know the Git metadata.

If you're deploying with the CLI or API, you'll need to set a new environment variable yourself.

**Reply by Vadorequest** [reply]
Yeah, this isn't told in the doc, it's what I figured out by experience, thanks for confirming.

**Reply by styfle** [reply]
It is [mentioned in the docs](https://vercel.com/docs/environment-variables#system-environment-variables) but its easy to miss (emphasis added)

> NOTE: System Environment Variables will be populated with an empty string if the value is not applicable, for example when running vercel dev during Development **or if the Deployment is not triggered by a Git Commit**.

We'll revise the docs to make sure this is more clear, thanks!

**Reply by Vadorequest** [reply]
Thanks! I suggest it'd also be useful in the Vercel UI (as shown above) when the option is usable. It'd make sense to show it there, **especially if the git integration isn't enabled for the project**!

**Reply by staff0rd** [reply]
How do you access this UI?

**Reply by hnykda** [reply]
I haven't found that UI above, but the vars are prefixed and accessible on vercel https://vercel.com/docs/projects/environment-variables/system-environment-variables#using-prefixed-framework-environment-variables-locally

**Comment by jokull** [comment]
Just an FYI, you can add it as a header for debugging or reloading for new bundle releases quite easily:

```js
  /* ... next.config.js */
  async headeres() {
    return [
      {
        source: '/*',
        headers: [{ key: 'Web-Build', value: process.env.VERCEL_GIT_COMMIT_SHA }]
      }
    ];
  },
```
